### PR TITLE
Add TRT support for ERNIE/BERT

### PR DESCRIPTION
**Pull request by cryoco**
added TRT support for ERNIE/BERT.
1、added trt op converter and plugin kernel for gelu activation op.
2、added trt op converter and plugin for layernorm op.
3、added op converter for fused multihead_matmul op.

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    auto* softmax_layer = TRT_ENGINE_ADD_LAYER(
+        engine_, SoftMax, *(qk_eltadd_layer->getOutput(0)));
+    auto* qkv_matmul_layer = TRT_ENGINE_ADD_LAYER(
+        engine_, MatrixMultiply, *(softmax_layer->getOutput(0)),
+        nvinfer1::MatrixOperation::kNONE, *(softmax_layer->getOutput(0)),
+        nvinfer1::MatrixOperation::kNONE);
+    auto* qkv_transpose_reshape_layer = TRT_ENGINE_ADD_LAYER(
+        engine_, Shuffle, *(qkv_matmul_layer->getOutput(0)));
+
+    nvinfer1::Dims3 qkv_reshape_dim(0, seq_len, head_number * size_per_head);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc]: ```qkv_reshape_dim(0, seq_len, head_number * size_per_head) ``` why 0 for first dimension?

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    qkv_transpose_reshape_layer->setFirstTranspose({0, 2, 1, 3});
+    qkv_transpose_reshape_layer->setReshapeDimensions(qkv_reshape_dim);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc:193]: 下一个层应该是mul，它只接受NCHW的输入，把qkv_reshape_dim最后一维度扩展1，或者对mul op前进行处理以接受三维度输入。

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    q_transpose_reshape_layer->setReshapeDimensions(head_reshape_dim);
+    q_transpose_reshape_layer->setSecondTranspose({0, 2, 1, 3});
+    k_transpose_reshape_layer->setReshapeDimensions(head_reshape_dim);
+    k_transpose_reshape_layer->setSecondTranspose({0, 2, 1, 3});
+
+    auto* k_scale_layer = TRT_ENGINE_ADD_LAYER(
+        engine_, Scale, *(k_transpose_reshape_layer->getOutput(0)),
+        nvinfer1::ScaleMode::kUNIFORM, shift.get(), scale.get(), power.get());
+
+    auto* qk_matmul_layer = TRT_ENGINE_ADD_LAYER(
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc:177]: 该层要设置输入是否要转置。

File: paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    int d1 = 0;
+    int d2 = 0;
+    int d3 = 0;
+    if (input_dims == 3) {
+      d0 = input_shape.d[0];
+      d1 = input_shape.d[1];
+      d2 = input_shape.d[2];
+      d3 = 1;
+    }
+    nvinfer1::Dims4 reshape_dim(d0, d1, d2, d3);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc]: 如果 input_dims !== 3, reshape_dim 会是{0, 0, 0, 0}？

File: paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    auto* reshape_layer = TRT_ENGINE_ADD_LAYER(
+        engine_, Shuffle, *const_cast<nvinfer1::ITensor*>(X));
+    reshape_layer->setReshapeDimensions(reshape_dim);
+    auto* input_after_reshape = reshape_layer->getOutput(0);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc]: 如果layer norm是plugin实现的话， 应该不需要进行reshape

File: paddle/fluid/inference/tensorrt/convert/gelu_op.cc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+class GeluOpConverter : public OpConverter {
+ public:
+  void operator()(const framework::proto::OpDesc& op,
+                  const framework::Scope& scope, bool test_mode) override {
+    VLOG(4) << "convert fluid gelu op to tensorrt gelu layer";
+
+    framework::OpDesc op_desc(op, nullptr);
+    // Declare inputs
+    int input_num = op_desc.Input("X").size();
+    PADDLE_ENFORCE(input_num == 1);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/gelu_op.cc]: 避免用PADDLE_ENFOCE

File: paddle/fluid/inference/tensorrt/convert/op_converter.h [hunk 7]
```diff
@@ -52,9 +52,11 @@ nvinfer1::Dims Vec2TRT_Dims(const std::vector<int64_t>& shape) {
                     "TensorRT' tensor input requires at least 2 dimensions");
   PADDLE_ENFORCE_LE(shape.size(), 4UL,
                     "TensorRT' tensor input requires at most 4 dimensions");
-  PADDLE_ENFORCE(shape.size() == 4UL || shape.size() == 2UL);
+  // PADDLE_ENFORCE(shape.size() == 4UL || shape.size() == 2UL);
```
Comment (NHZlX) [inline paddle/fluid/inference/tensorrt/convert/op_converter.h]: delete this

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc (hunk 1 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc]: Value 0 copies the corresponding dimension(batch size in this case) from input.
https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_shuffle_layer.html#a5fbca804e6da5b4cf89795ea83909fb6

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc (hunk 3 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc:177]: WIP.

File: paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc (hunk 4 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc]: if input_dims !== 3, there is no need to reshape, therefore we just copy the dims of input.

File: paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc (hunk 2 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc:193]: WIP.

File: paddle/fluid/inference/tensorrt/convert/gelu_op.cc (hunk 6 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/gelu_op.cc]: done.

File: paddle/fluid/inference/tensorrt/convert/op_converter.h (hunk 7 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/op_converter.h]: done.

File: paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc (hunk 5 above)
Comment (cryoco) [reply paddle/fluid/inference/tensorrt/convert/layer_norm_op.cc]: ok, WIP.

**Review by NHZlX — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LG const_cast

**Pull request merged**